karapace/serialization.py

```diff
@@ -0,0 +1,166 @@
+        result = await self.client.post("subjects/%s/versions" % quote(subject), json=payload)
+        if result.status_code != 200:
+            raise SchemaRetrievalError(result.json())
+        return result.json()["id"]
+
+    async def get_latest_schema(self, subject):
+        result = await self.client.get("subjects/%s/versions/latest" % quote(subject))
+        if result.status_code != 200:
+            raise SchemaRetrievalError(result.json())
+        json_result = result.json()
+        if "id" not in json_result or "schema" not in json_result:
+            raise SchemaRetrievalError("Invalid result format: %r" % json_result)
+        try:
+            return json_result["id"], avro.io.schema.parse(json_result["schema"])
+        except avro.io.schema.SchemaParseException as e:
+            raise SchemaRetrievalError("Failed to parse schema string from response: %r" % json_result) from e
+
+    async def get_schema_for_id(self, schema_id):
+        result = await self.client.get("schemas/ids/%d" % schema_id)
+        if result.status_code != 200:
+            raise SchemaRetrievalError(result.json()["message"])
+        json_result = result.json()
+        if "schema" not in json_result:
+            raise SchemaRetrievalError("Invalid result format: %r" % json_result)
+        try:
+            return avro.io.schema.parse(json_result["schema"])
+        except avro.io.schema.SchemaParseException as e:
+            raise SchemaRetrievalError("Failed to parse schema string from response: %r" % json_result) from e
+
+
+class SchemaRegistrySerializerDeserializer:
+    def __init__(self, **config):
+        if not config.get("config_path"):
+            raise Exception("config_path is empty or missing")
+        config_path = config["config_path"]
+        self.config = read_config(config_path)
+        registry_url = "%s:%d" % (self.config["host"], self.config["port"])
+        registry_client = SchemaRegistryClient(registry_url)
+        self.subject_name_strategy = topic_name_strategy
+        self.registry_client = registry_client
+        self.subjects_to_schemas = {}
+        self.ids_to_schemas = {}
+        self.schemas_to_ids = {}
+
+    @staticmethod
+    def serialize_schema(schema):
+        return json_encode(schema.to_json())
+
+    async def get_schema_for_topic(self, subject):
+        # apparently we can't really implement the other 2 and keep in line with the current kafka client api
+        # due to the serialize / deserialize method signature only passing topic as a relevant indicator
+        if subject in self.subjects_to_schemas:
+            return self.subjects_to_schemas[subject]
+        log.info("schema for subject %r not found locally, attempting to retrieve latest value", subject)
+        schema_id, schema = await self.registry_client.get_latest_schema(subject)
+        self.subjects_to_schemas[subject] = schema
+        self.schemas_to_ids[self.serialize_schema(schema)] = schema_id
```

Comment:
Should we consider using a hash as the key?

Reply:
It's probably a toss up between log visibility (no __repr__ method yet but one would probably include those 2 dicts), and CPU vs memory consumption . Unless memory does turn out to be an issue i'd rather leave it like this and only put it behind getters / setters otherwise